038-optimization-proposal-autoapproval.md

```diff
@@ -0,0 +1,66 @@
+# Auto-approval mechanism for optimization proposals
+
+This proposal is about allowing the user to create a `KafkaRebalance` custom resource for getting an optimization proposal and running a rebalance, but without the need for the manual approval via the `strimzi.io/rebalance=approve` annotation.
+This means that the user can simply create the `KafkaRebalance` custom resource and the corresponding generated optimization proposal will be approved automatically.
+
+## Current situation
+
+Currently, when the users want to run a cluster rebalance, they have to create a `KafkaRebalance` custom resource in order to generate an optimization proposal first.
+After getting the proposal, the only way to start the rebalancing operation is by approving it by annotating the custom resource with the `strimzi.io/rebalance=approve` annotation.
+There is the need for a manual intervention of the user and this is a two steps process.
+
+## Motivation
+
+In some cases, the manual approval process of the rebalance proposal is not necessary.
+Automatically approving an optimization proposal can save time from an operational point of view.
+It enables more automation where just creating a `KafkaRebalance` custom resource can go straight to the cluster rebalance without any additional manual interaction.
+
+## Proposal
+
+The `KafkaRebalance` custom resource can be extended with a new `spec.autoApproval` field for this purpose.
+
+### Auto-approval
+
+The auto-approval can be enabled just adding an empty `spec.autoApproval` field.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: # any mode
+  autoApproval: {}
```

Comment:
Do we know that (in the future when this is acting as more than an unconditional "approve it automatically" flag) it would actually be an _object_, rather than a _list_ of criteria to be met? That is, I could imagine that in that future we'd actually want something like

```yaml
spec:
  mode: ...
  approvalCriteria:
    - criterionType: Foo
      threshold: 100
    - criterionType: Bar
      someCriterionParameter: true
```

I _guess_ having in intermediate object might allow some flexibility (e.g. you could have an `anyOf` property and an `allOf` property to provide different ways of combining multiple criteria), but it's equally not clear to me that we'd need such flexibility (and hence no need for an intermediate object either).

Reply:
Note that `approvalCriteria` could allow use to handle the existing manual approval in a more consistent way. I.e. rather than using an annotation we could just say the API was to update the `spec.approvalCriteria` with a criterion that always evaluated as true. For example:

```yaml
spec:
  mode: ...
  # approvalCriteria: abence of property means not approved
```

And this CR

```yaml
spec:
  mode: ...
  approvalCriteria: 
    - criterionType: Always 
```

could be used for either:
* auto approval (if it was created like this) 
* manually approval (if edited to include this after the proposal was generated)

Reply:
I think that having a field named `approvalCriteria` right now when we don't have support for criteria is totally misleading for users. We have the desire to have criteria and rules but we don't have a real plan now, how much it's really feasible and if it will be really useful for the community so overcommitting could not help.
With the current `autoApproval` we could have a `criteria` or `rules` subfield to match your expectation. Maybe something like this?

```yaml
autoApproval:
  criteria:
    - type: Foo
      threshold: 100
    - type: Bar
      someCriterionParameter: true
```

which anyway leaves space in the `autoApproval` field to not having just a list.

Reply:
So, what about using annotation instead of changing the CRDs until we know more about the future CRD structure? E.g. if `strimzi.io/rebalance-auto-approval` is set to `true`, it is auto approved?

Reply:
We can go with annotation but I still don't get what makes `approvalCriteria` as list better than a generic `autoApproval` containing a field for the criteria list so allowing space for improvements. My feeling is that it's possible that we won't come back to have criteria anymore and we'll leave the annotation forever, so having an API (I talk about the `KafkaRebalance` CR) which mixes annotations and fields to provide features.

Reply:
> My feeling is that it's possible that we won't come back to have criteria anymore and we'll leave the annotation forever, so having an API (I talk about the KafkaRebalance CR) which mixes annotations and fields to provide features.

Well, if we see that we don't get back to it and don't need the criteria, we can convert the annotation into a boolean flag inside the KafkaRebalance CR the future. The end result in that case will be better than the empty object.

I approved the original design, so I was fine with it. Just trying to find some compromise to move forward with this. ;-)

Reply:
I will do the change on the proposal from field to annotation (I see @tombentley thumbs up on this idea) and moving the current one as rejected. Should we just track somewhere (an issue? a card on the project?) that we have to come back to this in the future to evaluate if promoting to field is feasible?

Reply:
> so having an API which (...) mixes annotations and fields to provide features

Well we already have that because we use an annotation for the approval anyway. If we really didn't come back to add support for criteria then leaving people with an API where the have to specify an empty object where they logically mean a boolean is pretty unintuitive. So on balance I think I prefer using an annotation for the time being until we know enough to design the API to support criteria properly.

Reply:
> Well we already have that because we use an annotation for the approval anyway

I see the use case different tbh. We are using annotation to take actions (approving, refreshing, ...) not for enabling or describing features. I see that the annotation usage fits pretty well when it's about "asking to the operator" to take an action on the CR. But of course, it could be just my view. I see the usage of annotation for configurability just as a temporary solution.

Reply:
The proposal is now updated to use an annotation.